src/main/java/com/cloudant/search3/Search.java

```diff
@@ -278,62 +280,58 @@ public void close() {
         } catch (final InterruptedException e) {
             // Ignored.
         }
-        handlers.forEach((index, handler) -> {
-            try {
-                handler.close();
-            } catch (final IOException e) {
-                LOGGER.catching(e);
-            }
-        });
-        handlers.clear();
+        synchronized(handlers) {
+            handlers.forEach((index, handler) -> {
+                try {
+                    handler.close();
+                } catch (final IOException e) {
+                    LOGGER.warn("Error while closing handler " + handler, e);
+                }
+            });
+            handlers.clear();
+        }
     }
 
     private <T> void execute(
             final Index index,
             final StreamObserver<T> responseObserver,
             final LuceneConsumer<SearchHandler> f) {
-        final SearchHandler handler = getOrOpen(index);
         try {
+            final SearchHandler handler = getOrOpen(index);
             f.accept(handler);
-            return;
         } catch (final IOException | AlreadyClosedException e) {
             failedHandler(index, e);
             responseObserver.onError(fromThrowable(e));
-            return;
         } catch (final ParseException e) {
             responseObserver.onError(fromThrowable(e));
-            return;
         } catch (final SessionMismatchException e) {
             responseObserver.onError(fromThrowable(e));
-            return;
         } catch (final RuntimeException e) {
             LOGGER.catching(e);
             responseObserver.onError(fromThrowable(e));
-            return;
         }
     }
 
-    private SearchHandler getOrOpen(final Index index) {
+    private SearchHandler getOrOpen(final Index index) throws IOException {
         final Subspace subspace = toSubspace(index);
         final Analyzer analyzer = SupportedAnalyzers.createAnalyzer(index);
 
-        final SearchHandler handler = handlers.computeIfAbsent(subspace, key -> {
-            try {
-                final SearchHandler result = searchHandlerFactory.open(db, subspace, analyzer);
-                scheduler.scheduleWithFixedDelay(
-                        new CommitOrCloseTask(subspace, result),
-                        commitIntervalSecs,
-                        commitIntervalSecs,
-                        TimeUnit.SECONDS);
-                LOGGER.info("Opened index {}", result);
+        synchronized(handlers) {
+            SearchHandler result = handlers.get(subspace);
+            if (result != null) {
                 return result;
-            } catch (final IOException e) {
-                LOGGER.catching(e);
-                return null;
```

Comment:
We no longer get an NPE. IIUC, you're letting the IOException go up the stack and where it is caught here:

https://github.com/cloudant-labs/search3-java/blob/9569ca2f746c152ec359574622480dda858eb443/src/main/java/com/cloudant/search3/Search.java#L302

Then in failedHandler you remove the index. Hence we see:

```
2019-11-20 22:01:19 WARN  Search:376 - Ignoring exception thrown while closing failed handler
java.io.FileNotFoundException: write.lock does not exist
	at com.cloudant.fdblucene.FDBDirectory.deleteFile(FDBDirectory.java:283) ~[search3-1.0.0-SNAPSHOT.jar:?]
	at com.cloudant.fdblucene.FDBLock.close(FDBLock.java:62) ~[search3-1.0.0-SNAPSHOT.jar:?]
	at org.apache.lucene.util.IOUtils.close(IOUtils.java:89) ~[search3-1.0.0-SNAPSHOT.jar:?]
	at org.apache.lucene.util.IOUtils.close(IOUtils.java:77) ~[search3-1.0.0-SNAPSHOT.jar:?]
	at org.apache.lucene.index.IndexWriter.rollbackInternalNoCommit(IndexWriter.java:2337) ~[search3-1.0.0-SNAPSHOT.jar:?]
	at org.apache.lucene.index.IndexWriter.rollbackInternal(IndexWriter.java:2268) ~[search3-1.0.0-SNAPSHOT.jar:?]
	at org.apache.lucene.index.IndexWriter.rollback(IndexWriter.java:2261) ~[search3-1.0.0-SNAPSHOT.jar:?]
	at com.cloudant.search3.FDBDirectorySearchHandler.close(FDBDirectorySearchHandler.java:82) ~[search3-1.0.0-SNAPSHOT.jar:?]
	at com.cloudant.search3.Search.failedHandler(Search.java:372) [search3-1.0.0-SNAPSHOT.jar:?]
	at com.cloudant.search3.Search.failedHandler(Search.java:362) [search3-1.0.0-SNAPSHOT.jar:?]
	at com.cloudant.search3.Search.execute(Search.java:303) [search3-1.0.0-SNAPSHOT.jar:?]
	at com.cloudant.search3.Search.updateDocument(Search.java:227) [search3-1.0.0-SNAPSHOT.jar:?]
	at com.cloudant.search3.grpc.SearchGrpc$MethodHandlers.invoke(SearchGrpc.java:763) [search3-1.0.0-SNAPSHOT.jar:?]
	at io.grpc.stub.ServerCalls$UnaryServerCallHandler$UnaryServerCallListener.onHalfClose(ServerCalls.java:171) [search3-1.0.0-SNAPSHOT.jar:?]
	at io.grpc.internal.ServerCallImpl$ServerStreamListenerImpl.halfClosed(ServerCallImpl.java:322) [search3-1.0.0-SNAPSHOT.jar:?]
	at io.grpc.internal.ServerImpl$JumpToApplicationThreadServerStreamListener$1HalfClosed.runInContext(ServerImpl.java:762) [search3-1.0.0-SNAPSHOT.jar:?]
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37) [search3-1.0.0-SNAPSHOT.jar:?]
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123) [search3-1.0.0-SNAPSHOT.jar:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_152]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_152]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_152]
```

Reply:
That's right. This PR is simplifying how errors are handled and logged, it doesn't fix anything per se. Hopefully it makes it easier for us to see how to fix things, though.